torch/csrc/jit/python/script_init.cpp

```diff
@@ -1094,21 +1110,72 @@ void initJitScriptBindings(PyObject* module) {
 
   py::class_<CompilationUnit, std::shared_ptr<CompilationUnit>>(
       m, "CompilationUnit")
-      .def(py::init<>())
+      .def(
+          py::init([](const std::string& lang, const uint32_t _frames_up) {
+            auto cu = std::make_shared<CompilationUnit>();
+            if (lang.size() > 0) {
+              compilationUnit_py_define(*cu, lang, nullptr, _frames_up);
+            }
+            return cu;
+          }),
+          py::arg("lang") = "",
+          py::arg("_frames_up") = 0)
+
       .def(
           "find_function",
           [](std::shared_ptr<CompilationUnit> self, const std::string& name) {
-            auto& fn = self->get_function(QualifiedName(name));
-            return StrongFunctionPtr(std::move(self), &fn);
+            auto fn = self->find_function(QualifiedName(name));
+            if (fn) {
+              return c10::optional<StrongFunctionPtr>(
+                  StrongFunctionPtr(std::move(self), fn));
+            } else {
+              return c10::optional<StrongFunctionPtr>(c10::nullopt);
+            }
+          })
+      .def(
+          "__getattr__",
+          [](std::shared_ptr<CompilationUnit> self, const std::string& name) {
+            auto fn = self->find_function(QualifiedName(name));
+            if (fn) {
+              return StrongFunctionPtr(std::move(self), fn);
+            } else {
+              throw AttributeError(
+                  "'CompilationUnit' has no attribute '%s'", name.c_str());
+            }
+          })
+      .def(
+          "get_functions",
+          [](const std::shared_ptr<CompilationUnit>& self) {
+            auto raw_functions = self->get_functions();
+            std::vector<StrongFunctionPtr> functions;
+            functions.reserve(raw_functions.size());
+            for (auto fn : raw_functions) {
+              if (fn) {
+                functions.emplace_back(self, fn);
+              }
+            }
+            return functions;
```

Comment:
Consider using `fmap` for this. Also, under what circumstances is it possible for `fn` to be `nullptr`?

Reply:
How do I skip `nullptr` with `fmap`? So the C++ `get_functions` gives a vector of functions. But for the name lookup, the `CompilationUnit` also keeps a dictionary `name->index`. This makes deleting a function (e.g. because we're defining a new one) delete from the dictionary and nullptr in the vector.

Reply:
Hm, looks like there is a function `filter` defined in `aten/src/ATen/core/functional.h` alongside `fmap` that lets you supply a predicate `fn` to skip elements.

Reply:
But will the C++ compiler optimize that for me or will that create a vector of intermediates? I didn't really feel sure, so I went for the explicit route.

Reply:
```
template<typename F, typename T>
inline std::vector<T> filter(at::ArrayRef<T> inputs, const F& fn) {
  std::vector<T> r;
  r.reserve(inputs.size());
  for(auto & input : inputs) {
    if (fn(input)) {
      r.push_back(input);
    }
  }
  return r;
}
```
It's inlined so I think there won't be any issue with intermediates.

Reply:
So I think it doesn't matter in the larger scheme of things, but so if I take inlining as being roughly the same as copying the code into the function rather than a call, I'm still creating a vector of `Function*` with the filter and then creating a second vector of `StrongFunctionPtr` with the `fmap`. It might be that compilers are that smart today, but I'm not sure that I would expect that.

Reply:
Yeah, I guess `T` here would be `Function*` and not `StrongFunctionPtr`...

Reply:
Ok, let's keep it the way it is.